subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/DefaultTaskClassInfoStore.java

```diff
@@ -57,30 +61,115 @@ public TaskClassInfo getTaskClassInfo(Class<? extends Task> type) {
+                if (taskActionFactory instanceof AbstractIncrementalTaskActionFactory
+                    && foundIncrementalTaskActionFactory != null
+                    && method.getName().equals(foundIncrementalTaskActionFactory.getMethod().getName())
+                ) {
+                    // Let's try if we can decide on one
+                    AbstractIncrementalTaskActionFactory selectedTaskAction = selectIncrementalTaskAction(type, foundIncrementalTaskActionFactory, taskActionFactory, current, method);
+                    if (selectedTaskAction != null) {
+                        foundIncrementalTaskActionFactory = selectedTaskAction;
+                        continue;
+                    }
+                }
+                if (previousDeclaringClass == declaringClass) {
+                    throw new GradleException(String.format(
+                        "Cannot use @TaskAction annotation on multiple overloads of method %s.%s()",
+                        declaringClass.getSimpleName(), method.getName()
+                    ));
+                } else if (previousDeclaringClass != null) {
+                    continue;
+                }
                 if (taskActionFactory instanceof AbstractIncrementalTaskActionFactory) {
-                    if (incremental) {
+                    if (foundIncrementalTaskActionFactory != null) {
                         throw new GradleException(String.format("Cannot have multiple @TaskAction methods accepting an %s or %s parameter.", InputChanges.class.getSimpleName(), IncrementalTaskInputs.class.getSimpleName()));
                     }
-                    incremental = true;
+                    foundIncrementalTaskActionFactory = (AbstractIncrementalTaskActionFactory) taskActionFactory;
+                    continue;
                 }
                 taskActionFactoriesBuilder.add(taskActionFactory);
             }
         }
+        if (foundIncrementalTaskActionFactory != null) {
+            taskActionFactoriesBuilder.add(foundIncrementalTaskActionFactory);
+        }
 
         return new TaskClassInfo(taskActionFactoriesBuilder.build(), cacheable);
     }
 
+    /**
+     * Select between two incremental task action factories.
+     *
+     * The selection works like this:
+     * - If both task actions are of the same type (InputChanges or IncrementalTaskInputs), we select the one we already found.
+     *   This happens when a subclass overrides the task action method in the subclass, while both have TaskAction annotations.
+     * - If the InputChanges is not deprecated and the IncrementalTaskInputs method is, then we try to select one:
+     *   - If a subclass overrides the IncrementalTaskInputs method, we create a {@link BridgingIncrementalInputsTaskActionFactory} which passes the {@link InputChanges} object into the IncrementalTaskInputs method
+     *   - If no subclass overrides the IncrementalTaskInputs method, we use the InputChanges method directly.
+     */
```

Comment:
Shall we mention that we do this complicated dance for AGP 3.5? Shall we promise to remove it in Gradle 6.1?

Reply:
You mean Gradle 7.0?

Reply:
I think it's okay to remove it earlier than that if we don't make it an official feature. Our goal is to make AGP 3.5 work with 6.0. Once AGP 3.6 is out we can remove this in the next minor release. Maybe it will be 6.2 or 6.3, but I would not wait until 7.0.

Reply:
I added a comment.